yield by itself generates invalid cpp

For example:
```
function foo() {
    yield 1;
}

function main() {
    foo();
}
```

Generates:
```cpp
#include <lib.h>
static void foo();


static void foo()
{
    using _JaktCurrentFunctionReturnType = void;
({ Optional<i64> __jakt_var_0;     {
        __jakt_var_0 = static_cast<i64>(1LL); goto __jakt_label_0;
    }
__jakt_label_0:; __jakt_var_0.release_value(); })}

ErrorOr<int> JaktMain::main(Array<String>)
{
    using _JaktCurrentFunctionReturnType = ErrorOr<int>;
    {
        foo();
    }
    return 0;
}
```

Which produces this:
```
/home/lukew/Desktop/repos/jakt/build/a.cpp:11:50: error: expected ';' after expression
__jakt_label_0:; __jakt_var_0.release_value(); })}
                                                 ^
                                                 ;
/home/lukew/Desktop/repos/jakt/build/a.cpp:11:18: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
__jakt_label_0:; __jakt_var_0.release_value(); })}
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
```

--- a/selfhost/parser.jakt
+++ b/selfhost/parser.jakt
@@ -260,6 +260,15 @@ struct ParsedBlock {
         return None
     }
 
+    function find_yield_keyword_span(this) -> Span? {
+        for stmt in .stmts.iterator() {
+            if stmt is Yield(expr) {
+                return stmt.span()
+            }
+        }
+        return None
+    }
+
     function span(this, parser: Parser) throws -> Span? {
         mut start: usize? = None
         mut end: usize = 0

--- a/selfhost/typechecker.jakt
+++ b/selfhost/typechecker.jakt
@@ -2095,6 +2095,11 @@ struct Typechecker {
             safety_mode: SafetyMode::Safe
         )
 
+        if block.yielded_type.has_value() {
+            .error_with_hint("Functions are not allowed to yield values", parsed_function.block.find_yield_span()!,
+                            "You might want to return instead", parsed_function.block.find_yield_keyword_span()!)
+        }
+
         // Typecheck return type a second time to resolve generics
         function_return_type_id = .typecheck_typename(
             parsed_type: parsed_function.return_type

--- a/tests/typechecker/block_yield_function.jakt
+++ b/tests/typechecker/block_yield_function.jakt
@@ -1,0 +1,6 @@
+/// Expect:
+/// - error: "Functions are not allowed to yield values"
+
+function main() {
+    yield 1
+}

